In [88]:
import pandas as pd

In [89]:
data = pd.read_csv('energydata_complete.csv')

In [112]:
data.head(2)

date  Appliances  lights     T1       RH_1    T2     RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.7900   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.7225   

      T3   RH_3    T4  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  19.79  44.73  19.0  ...  17.033333  45.53  6.600000        733.5    92.0   
1  19.79  44.79  19.0  ...  17.066667  45.56  6.483333        733.6    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  

[2 rows x 29 columns]

# 12

In [124]:
x = data[['T2']]
y = data[['T6']]
from sklearn.linear_model import LinearRegression
linear_model_ = LinearRegression()
linear_model_.fit(x, y)

LinearRegression()

In [127]:
predicted_values = linear_model_.predict(x)
from sklearn.metrics import r2_score
r2_score_ = r2_score(y, predicted_values)
round(r2_score_, 2)

0.64

# 13

In [90]:
import numpy as np

In [104]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
scaler = MinMaxScaler()
data_df = data.drop(columns=['date', 'lights'])

normalised = pd.DataFrame(scaler.fit_transform(data_df), columns=data_df.columns)
target = normalised['Appliances']
normal = normalised.drop(columns=['Appliances'])

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(normal,target,
test_size=0.3, random_state=42)
linear_model = LinearRegression()

linear_model.fit(x_train, y_train)

predicted_values = linear_model.predict(x_test)
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, predicted_values)
print(round(mae, 2))


0.05


# 14

In [92]:
rss = np.sum(np.square(y_test - predicted_values))
print(round(rss,2))

45.34


# 15

In [93]:
from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
round(rmse, 3) 

0.088

# 16

In [94]:
from sklearn.metrics import r2_score
r2_score = r2_score(y_test, predicted_values)
round(r2_score, 2)

0.15

# 17

In [111]:
feature_weights = pd.DataFrame(linear_model.coef_, index=normal.columns ,columns=['feature weights'])
print(feature_weights.sort_values(by='feature weights'))

             feature weights
rv2            -2.219382e+11
RH_2           -4.567699e-01
T_out          -3.220071e-01
T2             -2.361966e-01
T9             -1.899221e-01
RH_8           -1.576066e-01
RH_out         -7.770744e-02
RH_7           -4.455795e-02
RH_9           -3.980120e-02
T5             -1.567787e-02
T1             -3.281051e-03
Press_mm_hg     6.853342e-03
T7              1.027182e-02
Visibility      1.229110e-02
RH_5            1.602009e-02
RH_4            2.636302e-02
T4              2.897011e-02
Windspeed       2.919409e-02
RH_6            3.808389e-02
RH_3            9.594562e-02
T8              1.020135e-01
Tdewpoint       1.178244e-01
T6              2.364873e-01
T3              2.907092e-01
RH_1            5.536741e-01
rv1             2.219382e+11


this show that the objective of RH_2,RH_1 is correct


# 18

In [113]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(x_train,y_train)

Ridge(alpha=0.4)

In [116]:
predicted_values = ridge_reg.predict(x_test)
rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
round(rmse, 3)

0.088

there is no change to the root mean squared error

# 19

In [117]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=0.001)

In [118]:
feature_weights = pd.DataFrame(lasso_reg.coef_, index=normal.columns ,columns=['feature weights'])
print(feature_weights.sort_values(by='feature weights'))

             feature weights
RH_out             -0.049557
RH_8               -0.000110
T1                  0.000000
Tdewpoint           0.000000
Visibility          0.000000
Press_mm_hg        -0.000000
T_out               0.000000
RH_9               -0.000000
T9                 -0.000000
T8                  0.000000
RH_7               -0.000000
rv1                -0.000000
T7                 -0.000000
T6                  0.000000
RH_5                0.000000
T5                 -0.000000
RH_4                0.000000
T4                 -0.000000
RH_3                0.000000
T3                  0.000000
RH_2               -0.000000
T2                  0.000000
RH_6               -0.000000
rv2                -0.000000
Windspeed           0.002912
RH_1                0.017880


there are four features with non-zore feature weights

# 20

In [119]:
predicted_values = lasso_reg.predict(x_test)
rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
round(rmse, 3)

0.094

therefor the new RSME with lasso regression is 0.094